docs/book/src/reference/testdata/watching/controller.go

```diff
@@ -0,0 +1,218 @@
+		log.Error(err, "unable to fetch ConfigDeployment")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	// +kubebuilder:docs-gen:collapse=Begin the Reconcile
+
+	// your logic here
+
+	var configMapVersion string
+	if configDeployment.Spec.ConfigMap != "" {
+		configMapName := configDeployment.Spec.ConfigMap
+		foundConfigMap := &corev1.ConfigMap{}
+		err := r.Get(context.TODO(), types.NamespacedName{Name: configMapName, Namespace: configDeployment.Namespace}, foundConfigMap)
+		if err != nil {
+			// If a configMap name is provided, then it must exist
+			// You will likely want to create an Event for the user to understand why their reconcile is failing.
+			return ctrl.Result{}, err
+		}
+
+		// Hash the data in some way, or just use the version of the Object
+		configMapVersion = foundConfigMap.ResourceVersion
+	}
+
+	// Logic here to add the configMapVersion as an annotation on your Deployment Pods.
+
+	/* */
+
+	return ctrl.Result{}, nil
+}
+// +kubebuilder:docs-gen:collapse=Finish the Reconcile
+
+/*
+Finally, we add this reconciler to the manager, so that it gets started
+when the manager is started.
+
+Since we create dependency Deployments during the reconcile, we can specify that the controller `Owns` Deployments.
+
+However the ConfigMaps that we want to watch are not owned by the ConfigDeployment object.
+Therefore we must specify a custom way of watching those objects.
+This watch logic is complex, so we have split it into a separate method.
+*/
+
+func (r *ConfigDeploymentReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	ctrlBuilder := ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.ConfigDeployment{}).
+		Owns(&kapps.Deployment{})
+
+	var err error
+	if ctrlBuilder, err = r.watchUserConfigMaps(mgr, ctrlBuilder); err != nil {
+		return err
+	}
+
+	return ctrlBuilder.Complete(r)
+}
+
+/*
+There are two necessary steps that will allow us to watch the `ConfigMaps` that are referenced by `ConfigDeployments`.
+Index the reference field and setup a watch.
+*/
+
+func (r *ConfigDeploymentReconciler) watchUserConfigMaps(mgr ctrl.Manager, ctrlBuilder *builder.Builder) (*builder.Builder, error) {
+	/*
+	Determine the path of the field in the ConfigDeployment CRD that we wish to use as the "object reference".
+	This will be used in both the indexing and watching.
+	*/
+	configMapField := ".spec.configMap"
+
+	/*
+	This field must be indexed by the manager, so that we will be able to lookup `ConfigDeployments` by a referenced `ConfigMap` name.
+	This will allow for quickly answer the question:
+	- If ConfigMap _x_ is updated, which ConfigDeployments are affected?
+	 */
+
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &appsv1.ConfigDeployment{}, configMapField, func(rawObj runtime.Object) []string {
```

Comment:
Why the GetFieldIndexer().IndexField is required in this example?

Reply:
The field indexer is necessary in order to allow for easy lookup based on the "configMap" field in the spec. This makes the EnqueueRequestsFromMapFunc fast and easy. Otherwise we would need to list all ConfigDeployments from the namespace, and iterate through them finding the ones that use the ConfigMap passed through in the mapping function.